fgpyo/samtools.py

```diff
@@ -0,0 +1,375 @@
+            setting. The first value of the tuple will be used as the positive strand indicator,
+            the second value will be used as the negative strand indicator.
+    """
+
+    mark_strand_str: str
+    if mark_strand == FaidxMarkStrand.Custom:
+        assert custom_mark_strand is not None, (
+            "Cannot use custom mark strand without providing the custom strand indicators to "
+            + "`custom_mark_string`"
+        )
+        mark_strand_str = f"{mark_strand.value},{custom_mark_strand[0]},{custom_mark_strand[1]}"
+    else:
+        mark_strand_str = mark_strand.value
+
+    args: List[str] = ["--length", str(length), "--mark-strand", mark_strand_str]
+
+    if output is not None:
+        args.extend(["--output", str(output)])
+
+    if continue_if_non_existent:
+        args.append("--continue")
+    if reverse_complement:
+        args.append("--reverse-complement")
+    if fastq:
+        args.append("--fastq")
+
+    if fai_idx is not None:
+        args.extend(["--fai-idx", str(fai_idx)])
+
+    if gzi_idx is not None:
+        args.extend(["--gzi-idx", str(gzi_idx)])
+
+    if region_file is not None:
+        args.extend(["--region-file", str(region_file)])
+
+    args.append(str(sequence_file))
+
+    if regions is not None:
+        args.extend(regions)
+
+    return _pysam_faidx(*args)
+
+
+@enum.unique
+class SamIndexType(enum.Enum):
+    BAI = "-b"
+    CSI = "-c"
+
+
+def index(
+    alignment_file: Union[Path, Sequence[Path]],
+    output: Optional[Path] = None,
+    threads: int = 0,
+    index_type: SamIndexType = SamIndexType.BAI,
+    csi_index_size: int = 14,
+) -> str:
+    """
+    Calls the `samtools` function `index` using the pysam samtools dispatcher.
+
+    Arguments will be formatted into the appropriate command-line call which will be invoked
+    using the pysam dispatcher.
+
+    Returns the stdout of the resulting call to pysam.index()
+
+    Args:
+        alignment_file: Path to the SAM/BAM/CRAM file to index.
+        output: Path to the file to write output. (Currently may only be used when exactly one
+            alignment file is being indexed.)
+        threads: Number of input/output compression threads to use in addition to main thread.
+        index_type: The type of index file to produce when indexing.
+        csi_index_size: Sets the minimum interval size of CSI indices to 2^INT.
+    """
+
+    if isinstance(alignment_file, Path):
+        alignment_file = [alignment_file]
+    if output is not None and len(alignment_file) != 1:
+        raise ValueError("Only one output may be specified, for indexing only one alignment file")
+
+    args = [] if len(alignment_file) == 1 else ["-M"]
+
+    if index_type != SamIndexType.BAI:
+        args.append(index_type.value)
+
+    if index_type == SamIndexType.CSI:
+        args.extend(["-m", str(csi_index_size)])
+    args.extend(["-@", str(threads)])
+    args.extend(f"{input_path}" for input_path in alignment_file)
```

Comment:
The formatting of when you add a newline whitespace between code vs choosing not to is changing throughout the PR. I'd recommend keeping the style consistent.